Allow using wxBitmapBundle with wxDataViewBitmapRenderer

vadz opened the pull request:
While looking at #22359, I've realized that we couldn't use `wxBitmapBundle` to show images in wxDVC, so I've fixed this and now the following diff
```diff
diff --git a/samples/dataview/dataview.cpp b/samples/dataview/dataview.cpp
index 4b2f1d933e..4057a43959 100644
--- a/samples/dataview/dataview.cpp
+++ b/samples/dataview/dataview.cpp
@@ -904,7 +904,7 @@ void MyFrame::BuildDataViewCtrl(wxPanel* parent, unsigned int nPanel,
             wxDataViewColumn* const colCheckIconText = new wxDataViewColumn
                 (
                      L"\u2714 + icon + text",
-                     new wxDataViewCheckIconTextRenderer(),
+                     new wxDataViewBitmapRenderer(), // also works with explicit "wxBitmap"
                      MyListModel::Col_ToggleIconText,
                      wxCOL_WIDTH_AUTOSIZE
                 );
diff --git a/samples/dataview/mymodels.cpp b/samples/dataview/mymodels.cpp
index d6d76c2414..5aa6f20388 100644
--- a/samples/dataview/mymodels.cpp
+++ b/samples/dataview/mymodels.cpp
@@ -472,6 +472,11 @@ void MyListModel::GetValueByRow( wxVariant &variant,
             break;

         case Col_ToggleIconText:
+            if ( row % 2 )
+                variant << m_icon[row % 2]; // .GetBitmap(wxSize(16, 16)); -- not needed any longer, but still works too
+            else
+                variant = wxNullVariant;
+            break;
             {
                 wxString text;
                 wxCheckBoxState state;
```
compiles and works, i.e. shows crisp and not scaled bitmaps in high DPI.

@vslavik If you could please test this with Poedit, it would be great.

@MaartenBent @csomor Your review would be welcome as always, TIA!

vadz committed:
Add wxBitmapBundle::Clear()

This is just a convenient helper for resetting the bundle contents which
seems more readable than assigning an empty bundle to it.

vadz committed:
Use a single wxBitmapBundle in wxDataViewBitmapRenderer

Instead of using either a wxBitmap or a wxIcon, always use the same
wxBitmapBundle object to store whatever we are rendering.

This slightly simplifies the code and prepares for further changes, but
nothing real changes yet.

vadz committed:
Add support for storing wxBitmapBundle in a wxVariant

Define wxBitmapBundleVariantData without using the standard macros that
only work for wxObject-derived classes, but using more or less what they
expand into.

This will allow using wxBitmapBundle with wxDataViewCustomRenderer
subclasses in the upcoming commit.

vadz committed:
Allow relaxing the check for variant type in wxDataViewRenderer

Instead of checking for the exact variant type match, call the new
IsCompatibleVariantType() virtual function, which still does the same
check by default, but can be overridden to allow other types as well if
they're accepted by the renderer.

This will be soon used to allow accepting both wxBitmap (which must
still be accepted for compatibility) and wxBitmapBundle (which is the
simplest way to support high DPI) in wxDataViewBitmapRenderer at once.

vadz committed:
Accept either bitmaps or icons in wxDataViewBitmapRenderer

Allow using either wxBitmap or wxIcon in wxDataViewBitmapRenderer
independently of the type the renderer was created with because this is
convenient and there is no real danger in allowing to mix and match
bitmaps and icons together.

Override the just added IsCompatibleVariantType() to implement this.

vadz committed:
Accept wxBitmapBundle in wxDataViewBitmapRenderer too

This allows returning the entire bundle from the model GetValue()
function to let the renderer itself to select the best matching bitmap
to use.

vadz committed:
Use wxBitmapBundle as default type in wxDataViewBitmapRenderer

The default type doesn't really matter after the previous commits, as
the renderer accepts both wxBitmap and wxBitmapBundle (and also wxIcon)
in any case anyhow, but it seems better to use the preferred type as the
default one.

Also make the documentation more useful, although an example is still
lacking.

vadz commented:
Apparently nobody feels brave enough to have a look at this, so I'm going to use the only reliably working method for getting comments on my changes: merge it and see if it breaks anything.

vslavik commented:
> Apparently nobody feels brave enough to have a look at this,

Sorry I wasn't able to accommodate you in the past 3 business days  :/

It *looks* fine to me. I wanted to actually test it, as you asked, which means updating Poedit's build to it first on macOS and Windows, and I just wasn't able to do it yet.

vslavik commented on include/wx/dvrenderers.h:
```diff
@@ -155,6 +155,15 @@ class WXDLLIMPEXP_CORE wxDataViewRendererBase: public wxObject
+    // this renderer: by default, just compare it with GetVariantType(), but
+    // can be overridden to accept other types that can be converted to the
+    // type needed by the renderer.
+    virtual bool IsCompatibleVariantType(const wxString& variantType) const
```
> But I think we really want to allow the former

Yes, definitely.

vadz commented:
> > Apparently nobody feels brave enough to have a look at this,
> 
> Sorry I wasn't able to accommodate you in the past 3 business days :/

Sorry, my comment was intended to be a light-hearted one, nothing more. But, also, IME any comments about the PRs tend to come either relatively quickly (1-2 days) or not at all, so I thought it was safe to assume that none were upcoming here. I'm glad to be proven wrong, thanks for looking at this!

> It _looks_ fine to me. I wanted to actually test it, as you asked, which means updating Poedit's build to it first on macOS and Windows, and I just wasn't able to do it yet.

Please let me know if you'd prefer me to wait until you can test it before I merge it. If you can, it would, of course, better to wait until you can check that there are no unexpected problems with this. But I'd definitely like to merge this one before 3.1.7 (supposed to be released on 2022-06-06).

Thanks again!

vadz committed:
Improve wxDataViewRenderer variant type documentation

Explain what it is in the ctor documentation and link to the new
IsCompatibleVariantType() in GetVariantType() docs.

MaartenBent commented on src/qt/dvrenderers.cpp:
```diff
@@ -54,6 +54,13 @@ bool wxDataViewBitmapRenderer::GetValue( wxVariant &value ) const
+bool
+wxDataViewBitmapRenderer::IsCompatibleVariantType(const wxString& variantType) const
+{
+    return variantType == wxS("wxBitmap")
```
Does it needs `wxBitmapBundle` like the other platforms?

vadz replied:
I guess... But then this seems to be completely unimplemented in wxQt anyhow, and `wxBitmapBundle` is not implemented for it neither, so it's a bit of a moot point anyhow.

MaartenBent reviewed (commented):
Tested provided example on Windows and it seems to work good.

vadz commented on include/wx/qt/dvrenderers.h:
```diff
@@ -34,14 +34,16 @@ class WXDLLIMPEXP_ADV wxDataViewTextRenderer: public wxDataViewRenderer
     bool SetValue( const wxVariant &value );
     bool GetValue( wxVariant &value ) const;
+
+    bool IsCompatibleVariantType(const wxString& variantType) const;
```
This would result in "inconsistent use of override" warnings as it's not used for the other methods of the same class, so omitting it looks like a lesser evil (of course, ideal would be to add it everywhere but considering this is nothing more than a stub anyhow, I didn't feel like doing it would be useful).

vadz commented:
@vslavik Do you still plan to test this relatively soon?

vslavik commented:
@vadz I was slowed down by `wxBitmapBundle` changes breaking, of all things, the one platform that did natively handle bundles in `wxBitmap` just fine previously, and *none* of bitmaps, including in wxDVC, work in Poedit on macOS. Figuring that out seemed rather more pressing (patches incoming now that I did).

Testing of this so far: macOS is fine. wxGTK scales bitmaps in the list differently for HiDPI (NN instead of bilinear - unclear which is more correct, I'm fine with this). Didn't test Windows yet.

vadz commented:
> @vadz I was slowed down by `wxBitmapBundle` changes breaking, of all things, the one platform that did natively handle bundles in `wxBitmap` just fine previously, and _none_ of bitmaps, including in wxDVC, work in Poedit on macOS.

Oops, this was definitely unexpected. But now I'm even more glad that you could test it!

> Figuring that out seemed rather more pressing (patches incoming now that I did).

Looking forward to them, I'd like to understand how did I break it this time...

> Testing of this so far: macOS is fine. wxGTK scales bitmaps in the list differently for HiDPI (NN instead of bilinear - unclear which is more correct, I'm fine with this). Didn't test Windows yet.

As per 84cb293e71 (Make wxBitmap::Rescale() less horrible for commonly used icons, 2022-02-23), I think NN is better (well, less worse). If macOS does bilinear, we're probably not going to (be able?) change it, however.

vslavik commented:
I confirm that I couldn't find anything wrong with *this* patchset in Poedit on either of the 3 platforms.

I created  #22449 for some other issues with `wxBitmapBundle` transition.

> As per https://github.com/wxWidgets/wxWidgets/commit/84cb293e712ba2db7a946a166a83e1d0fdec3297 ... I think NN is better (well, less worse).

I have somewhat different opinion: while NN can "preserve sharp edges", it can also eradicate said edges altogether if rounding goes the wrong way when scaling *down*. However, I was commenting on 2x upscaling, where its reasonable (and I frankly don't know what macOS does in absence of Retina versions or what GTK+ native does).

vadz commented:
Thanks again for testing this, I'm going to merge this now as this shouldn't be affected by the art provider-related changes that still remain to be done.
